[Fix #337][Fix #345][Fix #341] Static Site build scripts

glogiotatidis opened the pull request:
- ~~Fix links to listing page with filtering~~.  moved to #361

glogiotatidis commented:
Marking this one WIP waiting for tomorrow's discussion on how we will automate it. 

/cc @mozmeao/theconn

jgmize commented:
This PR removes our security header implementations as well, which reminds me that we will need to re-implement them for a static site, which isn't possible to do with s3 (or GCS) alone, but can be done if we add [Cloudflare](https://scotthelme.co.uk/security-headers-cloudflare-worker/) or [Cloudfront](https://medium.com/@tom.cook/edge-lambda-cloudfront-custom-headers-3d134a2c18a2) (but not [Google CDN](https://issuetracker.google.com/issues/36427250)).

pmac commented:
It's easy with [Netlify](https://www.netlify.com/docs/headers-and-basic-auth/)... just FYI :)

glogiotatidis commented:
> This PR removes our security header implementations as well, which reminds me that we will need to re-implement them for a static site, which isn't possible to do with s3 (or GCS) alone, but can be done if we add [Cloudflare](https://scotthelme.co.uk/security-headers-cloudflare-worker/) or [Cloudfront](https://medium.com/@tom.cook/edge-lambda-cloudfront-custom-headers-3d134a2c18a2) (but not [Google CDN](https://issuetracker.google.com/issues/36427250)).

Added #362 to track this

glogiotatidis commented:
> It's easy with [Netlify](https://www.netlify.com/docs/headers-and-basic-auth/)... just FYI :)

Would like to try Netlify for this.

glogiotatidis commented:
I created a new branch named `static` and changed it to be the target for this PR, so we can co-ordinate and fix the [Milestone Issues](https://github.com/mozmeao/lumbergh/milestone/1) without yet touching the Django production site. Added a few more issues based on what we discussed yesterday

This should be ready for someone to review, as a mostly generic approach of building a static version of the site. I would probably need to adjust `build.sh` when we decide on build method.

glogiotatidis commented on docker-compose-build.yml:
```diff
@@ -0,0 +1,14 @@
+      - DATABASE_URL=sqlite:///database.sqlite
+      - DEBUG=False
+      - ALLOWED_HOSTS=*
+      - SECRET_KEY=59114b6a-2858-4caf-8878-482a24ee9542
```
Yes these are fine. Thanks for checking!

glogiotatidis commented on bin/build.sh:
```diff
@@ -0,0 +1,14 @@
+docker-compose -f docker-compose-build.yml -p careers-${BRANCH_NAME:-dev}-${BUILD_NUMBER:-1} run web ./bin/mirror.sh
+
+# Sync to S3 bucket
+pushd website_copy
```
Not really, will fix

glogiotatidis commented:
OK This is updated to work with netlify. You can check the PR deployed at https://deploy-preview-359--careers-stage.netlify.com/

I has been very easy to run the `mirror.sh` script on netlify and I overall very much liked the workflow, the checks they run and the build speed.

I'm +1 for running this on netlify. We have a few things to take care after merging this like (#360 #361 #362)

@mozmeao/theconn thoughts / reviews?

metadave commented:
This looks great! Clicking around the `deploy/netlify` check above, the site appears to be working correctly. Are there any tests that would detect broken links in the rendered HTML?

Next steps would be to get security and billing squared away w/ Netlify.

jpetto commented on bin/mirror.sh:
```diff
@@ -0,0 +1,44 @@
+wget --header "Host: ${CAREERS_HOST:-careers.mozilla.org}" \
+     --reject-regex "(.*)\?(.*)|static/(.*)" \
+     --mirror \
+     -nH -p -E http://0.0.0.0:8000/ http://0.0.0.0:8000/contribute.json
```
Could use a comment on this command. Lots of flags used that I definitely don't have memorized. 😉

glogiotatidis replied:
Added!

jpetto commented on netlify-headers:
```diff
@@ -0,0 +1,8 @@
+  X-Frame-Options: DENY
+  X-XSS-Protection: 1; mode=block
+  X-Content-Type-Options: nosniff
+  Content-Security-Policy: font-src 'self' http://*.mozilla.net https://*.mozilla.net http://*.mozilla.org https://*.mozilla.org; img-src 'self' data: http://*.mozilla.net https://*.mozilla.net http://*.mozilla.org https://*.mozilla.org http://www.google-analytics.com https://www.google-analytics.com; style-src 'self' 'unsafe-inline' http://*.mozilla.org https://*.mozilla.org http://*.mozilla.net ...
```
Not a big deal, but would be nice if this list could be stored in a more readable format.

glogiotatidis replied:
Agreed but that's the way netlify wants that.

pmac replied:
We could generate this file, but that is more work and this stuff is unlikely to change very often.

jpetto commented:
FWIW, here are the WebPageTest and Observatory results for current vs Netlify:

(Spoiler alert - they're basically identical, with Netlify winning in performance.)

## Current
- [WebPageTest](https://www.webpagetest.org/result/190123_WW_67551acb0ac632624ccbf77eaaa4ff4c/)
- [Observatory](https://observatory.mozilla.org/analyze/careers.mozilla.org)

## Netlify
- [WebPageTest](https://www.webpagetest.org/result/190123_GP_3f6054f7909a4f0c319a9810cba56037/)
- [Observatory](https://observatory.mozilla.org/analyze/deploy-preview-359--careers-stage.netlify.com)

jpetto commented:
Oh, I also ran the Netlify site through a broken link checker and it found none. (Well, fine, it found one, but it was in a job posting, which doesn't count/matter here.)

glogiotatidis commented:
Added wget docs and an additional acceptance test. I'm going to merge this one to `static` branch and continue to work against this branch until we're ready to go live with this.

Thanks for your time all!
